datadog.go

```diff
@@ -0,0 +1,277 @@
+package datadog
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// The maximum content size per request is 5MB
+	maxContentSize = 5 * 1024 * 1024
+
+	// The maximum size for a single log is 256kB
+	maxLogSize = 256 * 1024
+
+	// The maximum amount of logs that can be sent in a single request is 1000
+	maxLogCount = 1000
+
+	basePath          = "/v1/input"
+	datadogHostUS     = "https://http-intake.logs.datadoghq.com"
+	datadogHostEU     = "https://http-intake.logs.datadoghq.eu"
+	apiKeyHeader      = "DD-API-KEY"
+	defaultMaxRetries = 5
+)
+
+var (
+	defaultLevel = logrus.InfoLevel
+)
+
+// These options if provided will be added as default tags to each log sent
+type GlobalTags struct {
+	Service     string
+	Environment string
+	Maintainer  string
+	Application string
+	Hostname    string
+}
+
+type DatadogHook struct {
+	ApiKey          string
+	Tags            *GlobalTags
+	MinLevel        logrus.Level
+	MaxRetry        int
+	datadogEndpoint string
+	entryC          chan logrus.Entry
+	errorC          chan error
+	ticker          *time.Ticker
+	formatter       logrus.Formatter
+	wg              sync.WaitGroup
+}
+
+type Options struct {
+	ApiKey              *string
+	MinimumLoggingLevel *logrus.Level
+}
+
+func getenvOrDefault(key string, defaultString string) string {
+	if val := os.Getenv(key); val != "" {
+		return val
+	}
+
+	return defaultString
+}
+
+func defaultErrorhandler(err error) {
+	logrus.Errorf("The datadog logger hook has encountered an error: %s", err.Error())
+}
+
+func New(options *Options) (*DatadogHook, error) {
+	if options == nil {
+		options = &Options{}
+	}
+
+	if options.MinimumLoggingLevel == nil {
+		options.MinimumLoggingLevel = &defaultLevel
+	}
+
+	globalTags := &GlobalTags{
+		Service:     getenvOrDefault("SERVICE", "unknown"),
+		Environment: getenvOrDefault("ENVIRONMENT", "unknown"),
+		Maintainer:  getenvOrDefault("MAINTAINER", "unknown"),
+		Application: getenvOrDefault("APPLICATION", "unknown"),
+		Hostname:    getenvOrDefault("HOST", "0.0.0.0"),
+	}
+
+	region := os.Getenv("DATADOG_REGION")
```

Comment:
I would prefer options be passed in the options argument, instead of as environment variables, having environment variables as the only way to configure a package feels wrong IMO. 

As a side note our internal package are slightly different, as they are more a convenience thing.

Reply:
If we did want to have an environment variable config thing, I would recommend making a wrapper for this package.

Reply:
Yeah I did have that thought, I'm happy to do that I think that makes the most sense in terms of removing the GFS specifics from here, keep this simple and then wrap it for our own use to include the convenience stuff we like.

Reply:
Updated